web3/_utils/contracts.py

```diff
@@ -143,82 +133,25 @@ def find_matching_event_abi(
-) -> ABIFunction:
-    args = args or tuple()
-    kwargs = kwargs or dict()
-    num_arguments = len(args) + len(kwargs)
-
-    if fn_identifier is FallbackFn:
-        return get_fallback_func_abi(abi)
-
-    if fn_identifier is ReceiveFn:
-        return get_receive_func_abi(abi)
-
-    if not is_text(fn_identifier):
-        raise Web3TypeError("Unsupported function identifier")
-
-    name_filter = functools.partial(filter_by_name, fn_identifier)
-    arg_count_filter = functools.partial(filter_by_argument_count, num_arguments)
-    encoding_filter = functools.partial(filter_by_encodability, abi_codec, args, kwargs)
-
-    function_candidates = pipe(abi, name_filter, arg_count_filter, encoding_filter)
-
-    if len(function_candidates) == 1:
-        return function_candidates[0]
-    else:
-        matching_identifiers = name_filter(abi)
-        matching_function_signatures = [
-            abi_to_signature(func) for func in matching_identifiers
-        ]
-
-        arg_count_matches = len(arg_count_filter(matching_identifiers))
-        encoding_matches = len(encoding_filter(matching_identifiers))
-
-        if arg_count_matches == 0:
-            diagnosis = (
-                "\nFunction invocation failed due to improper number of arguments."
-            )
-        elif encoding_matches == 0:
-            diagnosis = (
-                "\nFunction invocation failed due to no matching argument types."
-            )
-        elif encoding_matches > 1:
-            diagnosis = (
-                "\nAmbiguous argument encoding. "
-                "Provided arguments can be encoded to multiple functions "
-                "matching this call."
-            )
-
-        collapsed_args = extract_argument_types(args)
-        collapsed_kwargs = dict(
-            {(k, extract_argument_types([v])) for k, v in kwargs.items()}
-        )
-        message = (
-            f"\nCould not identify the intended function with name `{fn_identifier}`, "
-            f"positional arguments with type(s) `{collapsed_args}` and "
-            f"keyword arguments with type(s) `{collapsed_kwargs}`."
-            f"\nFound {len(matching_identifiers)} function(s) with "
-            f"the name `{fn_identifier}`: {matching_function_signatures}{diagnosis}"
-        )
-
-        raise Web3ValidationError(message)
-
-
 def encode_abi(
     w3: Union["AsyncWeb3", "Web3"],
-    abi: ABIFunction,
+    abi: ABIElement,
     arguments: Sequence[Any],
     data: Optional[HexStr] = None,
 ) -> HexStr:
-    argument_types = get_abi_input_types(abi)
-
-    if not check_if_arguments_can_be_encoded(abi, w3.codec, arguments, {}):
+    argument_types = []
+    try:
+        argument_types = get_abi_input_types(abi)
+    except ValueError:
+        # Use the default argument_types if the abi doesn't have inputs
+        pass
+
+    if not check_if_arguments_can_be_encoded(
+        abi,
+        arguments,
+        {},
+        w3.codec,
+    ):
```

Comment:
Moving around the order of the args is a breaking change. It's an internal method it looks like, so that's fine, but we should make sure that this doesn't break anywhere else that calls `check_if_arguments_can_be_encoded`